arch/cortex-m/src/nvic.rs

```diff
@@ -1,58 +1,122 @@
 //! Cortex-M NVIC
-
-use kernel::common::cells::VolatileCell;
+//!
+//! Most NVIC configuration is in the NVIC registers:
+//! <https://developer.arm.com/docs/100165/0201/nested-vectored-interrupt-controller/nvic-programmers-model/table-of-nvic-registers>
+//!
+//! Also part of the NVIC conceptually is the ICTR, which in older versions of
+//! the ARM ARM was listed in the "Summary of system control and ID registers
+//! not in the SCB" and newer ARM ARMs just file it in its own little private
+//! sub-section with the NVIC documentation. Seems a configuration register
+//! without a home, so we include it in the NVIC files as it's conceptually here.
+//! <https://developer.arm.com/docs/ddi0337/latest/nested-vectored-interrupt-controller/nvic-programmers-model/interrupt-controller-type-register-ictr>
+
+use kernel::common::registers::{register_bitfields, register_structs, ReadOnly, ReadWrite};
 use kernel::common::StaticRef;
 
-#[repr(C)]
-// Registers for the NVIC
-struct NvicRegisters {
-    // Interrupt set-enable
-    iser: [VolatileCell<u32>; 8],
-    _reserved1: [u32; 24],
-    // Interrupt clear-enable
-    icer: [VolatileCell<u32>; 8],
-    _reserved2: [u32; 24],
-    // Interrupt set-pending (and read pending state)
-    ispr: [VolatileCell<u32>; 8],
-    _reserved3: [VolatileCell<u32>; 24],
-    // Interrupt clear-pending (and read pending state)
-    icpr: [VolatileCell<u32>; 8],
+register_structs! {
+    /// NVIC Registers.
+    ///
+    /// Note this generic interface exposes all possible NVICs. Most cores will
+    /// not implement all NVIC_XXXX registers. If you need to find the number
+    /// of NVICs dynamically, consult `ICTR.INTLINESNUM`.
+    NvicRegisters {
+        /// Interrupt Controller Type Register
+        (0x004 => ictr: ReadOnly<u32, InterruptControllerType::Register>),
+
+        /// Interrupt Set-Enable Registers
+        (0x100 => iser: [ReadWrite<u32, NvicSetClear::Register>; 32]),
+
+        /// Interrupt Clear-Enable Registers
+        (0x180 => icer: [ReadWrite<u32, NvicSetClear::Register>; 32]),
+
+        /// Interrupt Set-Pending Registers
+        (0x200 => ispr: [ReadWrite<u32, NvicSetClear::Register>; 32]),
+
+        /// Interrupt Clear-Pending Registers
+        (0x280 => icpr: [ReadWrite<u32, NvicSetClear::Register>; 32]),
+
+        /// Interrupt Active Bit Registers
+        (0x300 => iabr: [ReadWrite<u32, NvicSetClear::Register>; 32]),
+
+        (0x380 => _reserved0),
+
+        /// Interrupt Priority Registers
+        (0x400 => ipr: [ReadWrite<u32, NvicInterruptPriority::Register>; 252]),
+
+        (0x7f0 => @END),
```

Comment:
This won't work like this. As mentioned in the documentation for `register_structs`, you need to explicitly specify the "reserved" paddings. Otherwise the macro just puts the fields next to each other, ignoring the offsets.

In fact, these **offsets are only used for tests**, hence the importance of merging https://github.com/tock/tock/pull/1393 to test this.

(Note: The rationale to mandate the reserved fields is that there's no way to generate a sequence of identifiers like `_reserved0`, `_reserved1`, etc. with Rust macros)